### PR TITLE
Implement async and void keyword completion for local functions

### LINKED ISSUE #8616: Local functions: Async modifier not recommended while declaring local function.
``` C#
       void F(string[] args)
       {
           $$async void Foo()
       }
```

Expected: Async keyword offered in completion list.
Actual: Async keyword not offered.

### LINKED ISSUE #8617: Local functions: void keyword not recommended while declaring local function.
``` C#
       void F(string[] args)
       {
           $$void Foo()
       }
```

Expected: void keyword offered in completion list.
Actual: void keyword not offered.

### PR DESCRIPTION
**Pull request by agocke**
This PR implements completion for the simple cases of async and void in
a local function context. What it misses cases where parsing incorrectly
considers bare modifiers to be local variable declaration statements
rather than local function declaration statements. This is tracked by
bug #14525.

Fixes #8616
Fixes #8617

### CODE REVIEW

**Comment by agocke** [comment]
For review @dotnet/roslyn-ide

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs [hunk 1]
```diff
@@ -317,5 +317,19 @@ class C
 }";
             await VerifyProviderCommitAsync(markupBeforeCommit, "return", expectedCodeAfterCommit, commitChar: ';', textTypedSoFar: "return");
         }
+
+        [Fact]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs]: Missing trait.

File: src/EditorFeatures/CSharpTest2/Recommendations/AsyncKeywordRecommenderTests.cs [hunk 2]
```diff
@@ -6,6 +6,7 @@
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Recommendations
 {
+    [Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/AsyncKeywordRecommenderTests.cs]: IDE doesn't really use class traits.  Can you keep this on the individual function?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AsyncKeywordRecommender.cs [hunk 3]
```diff
@@ -22,7 +22,8 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
             }
 
             return !context.TargetToken.IsKindOrHasMatchingText(SyntaxKind.PartialKeyword)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AsyncKeywordRecommender.cs]: boolean expression too confusing.  Please break into individual checks.

File: src/Workspaces/CSharp/Portable/Utilities/SyntaxKindSet.cs [hunk 4]
```diff
@@ -82,5 +88,6 @@ internal class SyntaxKindSet
         {
             SyntaxKind.StructDeclaration,
         };
+
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/SyntaxKindSet.cs]: remove

File: src/EditorFeatures/CSharpTest2/Recommendations/VoidKeywordRecommenderTests.cs [hunk 5]
```diff
@@ -642,5 +642,51 @@ public class VoidKeywordRecommenderTests : KeywordRecommenderTests
...
+class C
+{
+    void M()
+    {
+        $$
+    }
+}");
+        }
+
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/14525")]
```
Comment (agocke) [inline src/EditorFeatures/CSharpTest2/Recommendations/VoidKeywordRecommenderTests.cs]: It's explained inline in the change, along with in the bug.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/VoidKeywordRecommenderTests.cs]: so much of hte keyword recommending system works totally by looking for token patterns.  The idea being that generally teh parse tree is not in a good state, especially when you have random incomplete snippets of text during typing.  So i'm somewhat surprised we wouldn't be able to handle this case, or that we would be thrown off by the parser doing something like this :)

Comment (agocke) [inline src/EditorFeatures/CSharpTest2/Recommendations/VoidKeywordRecommenderTests.cs]: @CyrusNajmabadi What would you suggest then? :)

File: src/EditorFeatures/CSharpTest2/Recommendations/AsyncKeywordRecommenderTests.cs (hunk 2 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest2/Recommendations/AsyncKeywordRecommenderTests.cs]: I'd kinda prefer we start though -- the number of times we've made a mistake with test-level traits makes me distrust our ability to use them without an analyzer.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs [hunk 6]
```diff
@@ -317,5 +317,19 @@ class C
...
+        {
+            var markup = @"
+class C
+{
+    public void M()
+    {
+        $$
+    }
+}";
+            await VerifyItemExistsAsync(markup, "async");
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs]: The AsyncKeywordRecommenderTests are being updated, why the test here too?

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 7]
```diff
@@ -292,6 +292,53 @@ public static bool IsAttributeNameContext(this SyntaxTree syntaxTree, int positi
...
 
+        public static bool IsLocalFunctionDeclarationContext(
+            this SyntaxTree syntaxTree,
+            int position,
+            CancellationToken cancellationToken)
+        {
+            var leftToken = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var token = leftToken.GetPreviousTokenIfTouchingWord(position);
+
+            // Local functions are always valid in a stmt context
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Spelling: stmt

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 8]
```diff
@@ -292,6 +292,53 @@ public static bool IsAttributeNameContext(this SyntaxTree syntaxTree, int positi
...
+                {
+                    // second appearance of "async" not followed by modifier: treat as type
+                    if (syntaxTree.GetPrecedingModifiers(token.SpanStart, token, cancellationToken)
+                        .Any(x => x == SyntaxKind.AsyncKeyword))
+                    {
+                        return false;
+                    }
+
+                    // rule out async lambda
+                    if (token.GetAncestor<ExpressionSyntax>() != null)
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Be careful with any checks like this: can you do an async local function inside of a lambda?

Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Please add tests either way.

File: src/Features/CSharp/Portable/project.json [hunk 9]
```diff
@@ -1,9 +1,11 @@
 {
+  "supports": {},
```
Comment (jasonmalinowski) [inline src/Features/CSharp/Portable/project.json]: Can these changes be put into a separate commit or PR explaining why?

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 10]
```diff
@@ -292,6 +292,53 @@ public static bool IsAttributeNameContext(this SyntaxTree syntaxTree, int positi
...
+                return true;
+            }
+
+            // Also valid after certain modifiers
+            var validModifiers = SyntaxKindSet.LocalFunctionModifiers;
+
+            var modifierTokens = syntaxTree.GetPrecedingModifiers(
+                position, token, cancellationToken);
+
+            if (modifierTokens.IsSubsetOf(validModifiers))
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: I didn't see any negative tests for after other modifiers?

File: src/EditorFeatures/CSharpTest2/Recommendations/VoidKeywordRecommenderTests.cs (hunk 5 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest2/Recommendations/VoidKeywordRecommenderTests.cs]: @agocke I think @CyrusNajmabadi is suggesting you fix it.

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 7 above)
Comment (agocke) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Sms splld crrctly t m

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs (hunk 6 above)
Comment (agocke) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/KeywordCompletionProviderTests.cs]: Oh, yeah, I added that before I found the AsyncKeywordRecommenderTests. I'll remove it.

**Comment by agocke** [comment]
@CyrusNajmabadi @jasonmalinowski I think I've addressed all your comments

**Comment by CyrusNajmabadi** [comment]
Checking

File: src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs [hunk 11]
```diff
@@ -16,7 +16,14 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
     internal static partial class SyntaxTreeExtensions
     {
         public static ISet<SyntaxKind> GetPrecedingModifiers(
-            this SyntaxTree syntaxTree, int position, SyntaxToken tokenOnLeftOfPosition, CancellationToken cancellationToken)
+            this SyntaxTree syntaxTree,
+            int position,
+            SyntaxToken tokenOnLeftOfPosition,
+            CancellationToken cancellationToken) =>
+            syntaxTree.GetPrecedingModifiers(position, tokenOnLeftOfPosition, out var ignore);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs]: out _    :'(

**Comment by CyrusNajmabadi** [comment]
👍   Thanks!

File: src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs (hunk 11 above)
Comment (agocke) [inline src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs]: Not in the toolset compiler yet.

File: src/EditorFeatures/CSharpTest2/Recommendations/AsyncKeywordRecommenderTests.cs [hunk 12]
```diff
@@ -130,5 +131,113 @@ class Foo
     partial $$
 }");
         }
+
+        [Fact]
+        [WorkItem(8616, "https://github.com/dotnet/roslyn/issues/8616")]
+        [Test.Utilities.CompilerTrait(Test.Utilities.CompilerFeature.LocalFunctions)]
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest2/Recommendations/AsyncKeywordRecommenderTests.cs]: We have CompilerTraits as well? We didn't unify? :-(

File: src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs [hunk 13]
```diff
@@ -16,7 +16,14 @@ namespace Microsoft.CodeAnalysis.CSharp.Extensions
...
     {
         public static ISet<SyntaxKind> GetPrecedingModifiers(
-            this SyntaxTree syntaxTree, int position, SyntaxToken tokenOnLeftOfPosition, CancellationToken cancellationToken)
+            this SyntaxTree syntaxTree,
+            int position,
+            SyntaxToken tokenOnLeftOfPosition,
+            CancellationToken cancellationToken)
+            => syntaxTree.GetPrecedingModifiers(position, tokenOnLeftOfPosition, out var _);
+
+        public static ISet<SyntaxKind> GetPrecedingModifiers(
```
Comment (jasonmalinowski) [inline src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs]: Please document what "newPosition" is, or give a more descriptive name.
